### PR TITLE
Rework `HomeworkDialogBloc` .

### LINKED ISSUE #1115: Markdown analytics are wrong when editing a homework.
## Describe the bug
When editing a homework and not using markdown in the new description a "`markdown_used`" analytics event is logged regardless.

## Steps to reproduce the bug

1. Go to homework creation page
2. Create a homework with the description: "no markdown"
3. Edit the homework and change the description to "still no markdown"
4. Save edited homework.

## Current _broken_ behavior

A `markdown_used`-event is logged.

## Expected behavior
No `markdown_used`-event should be logged.

## Tested device

- App version: 1.7.9

### PR DESCRIPTION
**Pull request by Jonas-Sander**
Replaced `HomeworkDialogBloc` with a `Bloc` from [package:bloc](https://pub.dev/packages/bloc).  
Added unit and golden tests.

I tried to keep the behavior mostly the same. There are still some of the same bugs as before (e.g. #1117), they should be fixed in separate PRs.

Differences to old behavior:
* If any field is changed the quit prompt will appear (e.g. submission enabled, submission time and is private were ignored before)
* Pressing save will show errors (red text) like before, but not a `SnackBar`

Not so great:
* `PrefilledTextField` will we rebuild each time the text changes (title, description) - didn't see any problems because of this though
* Code quality is mediocre (e.g. using `HomeworkDto` for the current state in the bloc which lead to _horrible_ hacks for the todo `DateTime`)
* Quality of tests if mediocre, good amount of duplication. Also instead of testing each thing seperately I threw some different scenarios (private, not private, with submissions) together which just tested different things at the same time.
* Markdown analytics not tested automatically, just manually once
* No automatic tests for awaiting or not awaiting API calls depending on if attachments are added because of the Firestore offline behavior where Future won't complete (it works though, tested it manually several times)

Good:
* Good amount of test coverage
* Several golden tests (happy path and errors)
* UI is decoupled from business logic
* Should be a bit better to add functionality and maintain than before
* 🗿

https://github.com/SharezoneApp/sharezone-app/assets/29028262/86ef445f-2424-4607-b3fc-d403d70eafa2   

https://github.com/SharezoneApp/sharezone-app/assets/29028262/64332530-0181-4bbf-8a35-d31e62405bc8 

Fixes #1115

### CODE REVIEW

File: app/lib/pages/homework/homework_dialog.dart [hunk 1]
```diff
@@ -174,50 +175,85 @@ class __HomeworkDialogState extends State<__HomeworkDialog> {
...
+            ScaffoldMessenger.of(context).hideCurrentSnackBar();
+            showLeftRightAdaptiveDialog(
+              content: Text(
+                  'Hausaufgabe konnte nicht gespeichert werden.\n\n$error\n\nFalls der Fehler weiterhin auftritt, kontaktiere bitte den Support.'),
+              left: null,
+              right: AdaptiveDialogAction.ok,
+              context: context,
+            );
+        }
+      },
```
Comment (Jonas-Sander) [inline app/lib/pages/homework/homework_dialog.dart:196]: Used [`package:bloc_presentation`](https://pub.dev/packages/bloc_presentation) here.

**Comment by Jonas-Sander** [comment]
@nilsreichardt Ready to review. Tbh some of the code is a bit shitty, test code is not perfect (good amount of duplication) but my motivation is decreasing and I don't want this to be open for forever. Do feel free to critique, and suggest improvements though.

**Review by nilsreichardt — APPROVED** [review]
image

Is that correct that the due date is red when it's already filled out? As a user, when a due date is red, I would assume that something is wrong with the due date.

Is it still possible to display still a snack bar when there is an error? I know that it's now difficult to display a concrete error message, but we could still display something like "Bitte fülle alle erforderlichen Felder aus". This provides the user a clear action what to do (and it's more a11y friendly because the snack bar will automatically be read by the screen reader when it pops up).

LGTM

**Comment by Jonas-Sander** [comment]
> Is that correct that the due date is red when it's already filled out? As a user, when a due date is red, I would assume that something is wrong with the due date.

No that was a bug in the goldens test code setup that I now fixed.   

> Is it still possible to display still a snack bar when there is an error? I know that it's now difficult to display a concrete error message, but we could still display something like "Bitte fülle alle erforderlichen Felder aus". This provides the user a clear action what to do (and it's more a11y friendly because the snack bar will automatically be read by the screen reader when it pops up).
> 
> LGTM

I added a SnackBar that always says: "Bitte fülle alle erforderlichen Felder aus" if the save button is pressed and not all necessary fields are filled out 👍

**Pull request merged**